File: 07 July/07-10/exactChange.js
function exactChange(amount) {
/* Bottom-up approach using tabulation (iterative solution)
  const DENOMINATIONS = [1, 5, 10, 25];

  const distinctWays = Array.from({ length: amount + 1 }, (_, index) => index === 0 ? 1 : 0);

  for (const denomination of DENOMINATIONS) {
    for (let i = denomination; i <= amount; ++i) {
      distinctWays[i] += distinctWays[i - denomination];
    }
  }

  return distinctWays[amount];
*/

// Top-down approach using memoization (recursive solution)
  const DENOMINATIONS = [25, 10, 5, 1];
  const cache = new Map();

  const countWays = (amountRemaining, coinIndex) => {
    const key = `${amountRemaining}_${coinIndex}`;

    if (cache.has(key)) {
      return cache.get(key);
    }

    if (amountRemaining === 0) {
      cache.set(key, 1);
      return 1;
    }

    let distinctWaysCount = 0;

    for (let i = coinIndex; i < DENOMINATIONS.length; ++i) {
      if (DENOMINATIONS[i] <= amountRemaining) {
        distinctWaysCount += countWays(amountRemaining - DENOMINATIONS[i], i);
      }
    }

    cache.set(key, distinctWaysCount);

    return distinctWaysCount;
  }

  return countWays(amount, 0);
}

// --- TEST SUITE ---

const testsText = String.raw`
1. exactChange(3) should return 1.
2. exactChange(9) should return 2.
3. exactChange(17) should return 6.
4. exactChange(39) should return 24.
5. exactChange(61) should return 73.
6. exactChange(99) should return 213.
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("——————————————————————————",
            "\n🧪Starting Verification...",
            "\n——————————————————————————");

  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(`(${test.output})`);

    if (JSON.stringify(functionCallOutput) === JSON.stringify(testOutput)) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    } else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${JSON.stringify(testOutput)}\nGot: ${JSON.stringify(functionCallOutput)}`);

      ++failCount;
    }
    console.log("————————————————————————————");
  })

  console.log(failCount
    ? `⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`
    : "🎉SUCCESS: All tests PASSED."
  );
}

runTests(testData);

File: 07 July/07-10/exact_change.py
def exact_change(amount):
# Bottom-up approach using tabulation (iterative solution)
#    DENOMINATIONS = [1, 5, 10, 25]
#
#    distinct_ways = [1 if i == 0 else 0 for i in range(amount + 1)]
#
#    for denomination in DENOMINATIONS:
#        for i in range(denomination, amount + 1):
#            distinct_ways[i] += distinct_ways[i - denomination]
#
#    return distinct_ways[amount]

# Top-down approach using memoization (recursive solution)
    DENOMINATIONS = [25, 10, 5, 1]
    cache = {}

    def count_ways(amount_remaining, coin_index):
        key = amount_remaining, coin_index

        if key in cache:
            return cache[key]
        
        if amount_remaining == 0:
            cache[key] = 1
            return 1

        distinct_ways_count = 0

        for i in range(coin_index, len(DENOMINATIONS)):
            if DENOMINATIONS[i] <= amount_remaining:
                distinct_ways_count += count_ways(amount_remaining - DENOMINATIONS[i], i)

        cache[key] = distinct_ways_count

        return distinct_ways_count

    return count_ways(amount, 0)

# --- TEST SUITE ---

tests_text = r'''
1. exact_change(3) should return 1.
2. exact_change(9) should return 2.
3. exact_change(17) should return 6.
4. exact_change(39) should return 24.
5. exact_change(61) should return 73.
6. exact_change(99) should return 213.
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("——————————————————————————",
        "\n🧪 Starting Verification...",
        "\n——————————————————————————")

    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            fail_count += 1

        print("————————————————————————————")

    print(
        f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED." if fail_count
        else "🎉 SUCCESS: All tests PASSED."
    )

run_tests(test_data)